Handle headers from the signed part of unencrypted signed message (#3986)

iequidoo opened the pull request:
This makes DC compatible with "multipart/signed" messages thus allowing switching to them someday
from the current "multipart/mixed" unencrypted message format.

r10s commented:
> Note that this breaks user avatars for old clients

what does "old" mean here - all clients up to core 1.108?

if so, we should consider to split this pr into two parts: one that changes reading - and another once that changes writing, the latter is shipped some months later or so.

otherwise, we will easily have a period of several months where user avatars are not working - it takes quite some time until we have releases out on all platforms - and then it needs time until ppl install and use them.

iequidoo commented:
> what does "old" mean here - all clients up to core 1.108?

Unfortunately, yes.

> if so, we should consider to split this pr into two parts: one that changes reading - and another one that changes writing. the latter is merged some ui releases later then.
> 
> otherwise, we will easily have a period of several months where user avatars are not working reliably - it takes quite some time until we devs have releases out on all platforms - and then it needs time until ppl install and use them.

There is another option also (also a w/a for several months until users get updated) -- to attach avatars in the outer (not signed) message part. So, avatars won't be signed in unencrypted messages but this preserves the current behaviour. I think i will go this way

iequidoo commented:
> There is another option also (also a w/a for several months until users get updated) -- to attach avatars in the outer (not signed) message part. So, avatars won't be signed in unencrypted messages but this preserves the current behaviour. I think i will go this way

Oh no, current clients look for an avatar only in a "multipart/mixed" part. So, the only simple w/a (for several months also) is to omit attaching avatars to unencrypted messages at all. So, an avatar will be attached in the first encrypted message and then every 14 (`DC_RESEND_USER_AVATAR_DAYS`) days. But if we can't encrypt because it is, say, a group chat with ppl not using Autocrypt, there will be no avatar exchange. IDK how critical it is.

Another option is to wrap the "multipart/signed" part in an extra "multipart/mixed" part where an avatar will be attached. But it's ugly and the resulting message structure will differ from one of signed unencrypted messages created by other MUAs like TB.

@r10s @link2xt what do you think?

r10s commented:
once step back, first :) 

> Although it does a little for security, it will help to protect from unwanted server-side
modifications and bugs.

this means, all outgoing messages are now signed, right? 
is it really a thing that server modification takes place in the inner mime parts?
in our experience, even outlooks leaves that alone (that was the reason, we put the avatar there, and not to the outer mime part).

so question: was this discussed on a broader, general scale?
is it a change we really want?
i can at least imaging that this could cause troubles with other, maybe older mua, eg. showing weird attachments and do not support `multipart/signed` gracefully. that would worsening overall ux with delta chat. are there other mua out, that really always and by default sign all messages?
(for encryption we have a signal that the other side can somehow handle that - we do not have that for sign-only)

if we finally decide for sending everything as `multipart/signed`:

is would not scarify avatar ux at all, we spend quite some effot in make it as usuable as it is currently, wrt outlook, wrt using inner headers, wrt not showing the attachment to normal mua.
i think, the user avatar is a great thing, that is loved by many ppl - eg. one of the largest requests were never about security but adding an option to enlarget received avatars in ui :)

so, _if_ we go for it, we need to do it by two steps:
(1) read avatars from within multipart/signed and add tests for that, do releases on all ui
(2)  two releases or so later: switch to multipart/signed

good thing about that is that we can do a a pr for (1) already just now - all the questions wrt other mua and multipart/signed at all, can be discussed in the time between (1) and (2)

iequidoo commented:
> this means, most outgoing messages will be signed using `multipart/signed`, right?

Only all unencrypted messages, of course.

> is it really a thing that server-side modifications take place in the inner mime parts?

We experienced this in https://github.com/deltachat/deltachat-core-rust/issues/3986 . I couldn't reproduce it with Gmail f.e. and even with gmx.com, but @gerryfrancis reproduced it for me with gmx.at (for some reason i couldn't create an account at that domain). So, the `Content-Transfer-Encoding: quoted-printable` header and the extra line break are added by the server. And also it's not a problem in DC displaying the wrong line break because i see it in the message source and Thunderbird also displays it.

> and do the server really stop doing so just if there is a signature? from the view of the server it is still just a multipart/sth. in our experience, even outlooks leaves inner mime headers mostly alone (that was the reason, we put the avatar there, and not to the outer mime part).

IDK, but decided to try and see how it helps. I thought it's a good idea anyway because we can then verify all signed messages if we have verified the signing key at some point.

> so question: was this change discussed on a broader, general scale? is it a change we really want? i'd like to throw in this blob post of of the of the K-9 and OpenKeychain authors: https://k9mail.app/2016/11/24/OpenPGP-Considerations-Part-I (cite: "Signed-Only Mails Considered Harmful" :) it does not really match as we do not want to have ui around it, but still :)

I think we can (and already do) discuss it here :) And yes, i agree with the arguments in "OpenPGP Considerations, Part I: Signed-Only Mails" wrt complexity in UX and ecosystem (but after all, i don't think that signed-only mails are useless at all). So, go for "multipart/signed" only if it doesn't break compatibility with other MUAs. I believe most of them should handle it correctly, but that needs an investigation.

> apart from that, i can easily imagine that multipart/signed could cause troubles with other, maybe older mua, eg. showing weird attachments and do not support multipart/signed gracefully. that would worsening overall ux with delta chat. are there other mua out, that really always and by default sign all messages? (for encryption we have a signal that the other side can somehow handle that - we do not have that for sign-only)

IDK MUAs doing this by default but it can be set up in Thunderbird f.e. I even used this mode, but now i mostly use DC for mailing.
 
> _if_ we finally decide for sending everything as `multipart/signed`:
> 
> i would not scarify avatar ux at all, we spend quite some effort in making it as usuable as it is currently, wrt outlook, wrt using inner headers, wrt not showing the attachment to normal mua. i think, the user avatar is a great thing, that is loved by many ppl - eg. one of the largest requests were never about security but adding an option to enlarget received avatars in ui :)
> 
> so, if we go for multipart/signed, we need to do it in two steps: (1) read avatars from within multipart/signed and add tests for that, do releases on all ui (2) two releases or so later: switch to multipart/signed
> 
> good thing about that is that we can do a a pr for (1) already just now - all the questions wrt other mua and multipart/signed at all, can be discussed in the time between (1) and (2)

I agree, doing just (1) for now is reasonable. Will split the second part out of this PR. Then we have time to think and evaluate all pros and cons. Thanks for comments!

iequidoo commented:
@r10s, so, what's done:
- Added a support for "multipart/signed" message format
- Added a config option to make all messages with Autocrypt header "multipart/signed"
- Added a test `test_selfavatar_unencrypted_signed` for the "multipart/signed" format. In fact it checks not only the avatar header, but the whole message structure. I hope this test is sufficient, the more complex part is to test compatibility with other MUAs.

Note also that i don't add the `Subject` header outside of the signed message part as Thunderbird does f.e. to not complicate the code and the message structure. [RFC 3156](https://www.rfc-editor.org/rfc/rfc3156) doesn't require that. IDK why TB does that. If it is for compatibility with other MUAs, we probably should do the same

iequidoo commented on src/mimeparser.rs:
```diff
@@ -224,8 +224,33 @@ impl MimeMessage {
+                    (&mail, mimetype)
+                }
+            } else {
+                // Old clients do not sign unencrypted messages.
```
Changed to "Currently we do not sign unencrypted messages by default". Non-DC clients are unrelated here, this code block is for avatars only

iequidoo commented on src/mimeparser.rs:
```diff
@@ -224,8 +224,33 @@ impl MimeMessage {
+                    );
+                    (part, part.ctype.mimetype.parse::<Mime>()?)
+                } else {
+                    // If it's a partially fetched message, there are no subparts and this actually
```
I wanted to say that it's not an error and we should just skip all "multipart/mixed"-related stuff if so. But actually we jump to the check `if mimetype.type_() == mime::MULTIPART && mimetype.subtype().as_str() == "mixed"` from here although it's not needed. But rewriting the code to skip this check would make the code more complex, so i leaved it as is, anyway this check won't pass.

Removed the second part of the comment as it doesn't help to understand why the code is written this way. I think it is not so important.
